protobuf_serialization/writer.nim

```diff
@@ -0,0 +1,159 @@
+#Writes the specified type into a buffer using the Protobuf binary wire format.
+
+import options
+
+import stew/shims/macros
+import faststreams/outputs
+import serialization
+
+import internal
+import types
+
+proc newProtobufKey(number: int, wire: ProtobufWireType): seq[byte] =
+  result = newSeq[byte](10)
+  var viLen = 0
+  doAssert encodeVarInt(result, viLen, PInt((number shl 3) or int(wire))) == VarIntStatus.Success
+  result.setLen(viLen)
+
+proc writeProtobufKey(
+  stream: OutputStream,
+  number: int,
+  wire: ProtobufWireType
+) {.inline.} =
+  stream.write(newProtobufKey(number, wire))
+
+proc writeVarInt(stream: OutputStream, fieldNum: int, value: VarIntWrapped) =
+  let bytes = encodeVarInt(value)
+  if (bytes.len == 1) and (bytes[0] == 0):
+    return
+  stream.writeProtobufKey(fieldNum, VarInt)
+  stream.write(bytes)
+
+proc writeFixed(stream: OutputStream, fieldNum: int, value: auto) =
+  when sizeof(value) == 8:
+    let wire = Fixed64
+  else:
+    let wire = Fixed32
+  if value.unwrap() == 0:
+    return
+
+  stream.writeProtobufKey(fieldNum, wire)
+  stream.encodeFixed(value)
+
+#stdlib types toProtobuf's. inlined as it needs access to the writeValue function.
+include stdlib_writers
+
+proc writeValueInternal[T](stream: OutputStream, value: T)
+
+proc writeLengthDelimited[T](
+  stream: OutputStream,
+  fieldNum: int,
+  rootType: typedesc[T],
+  fieldName: static string,
+  flatValue: LengthDelimitedTypes
+) =
+  var cursor = stream.delayVarSizeWrite(10)
+  let startPos = stream.pos
+
+  #Byte seqs.
+  when flatValue is CastableLengthDelimitedTypes:
+    if flatValue.len == 0:
+      return
+    stream.write(cast[seq[byte]](flatValue))
+
+  #Standard lib types which use custom converters, instead of encoding the literal Nim representation.
+  elif type(flatValue).isStdlib():
+    stream.stdlibToProtobuf(rootType, fieldName, flatValue)
+
+  #Nested object which even if the sub-value is empty, should be encoded as long as it exists.
+  elif rootType.isPotentiallyNull():
+    writeValueInternal(stream, flatValue)
+
+  #Object which should only be encoded if it has data.
+  elif flatValue is (object or tuple):
+    writeValueInternal(stream, flatValue)
+
+  else:
+    {.fatal: "Tried to write a Length Delimited type which wasn't actually Length Delimited.".}
+
+  if (stream.pos != startPos) or (rootType.isPotentiallyNull()):
+    cursor.finalWrite(newProtobufKey(fieldNum, LengthDelimited) & encodeVarInt(PInt(int32(stream.pos - startPos))))
+  else:
+    cursor.finalWrite([])
```

Comment:
This code will write zero bytes in place of the length prefix (effectively omitting the prefix). I suspect this is not what you wanted to do. You probably wanted to write a single zero byte?

Reply:
If the length is 0, we don't encode the buffer at all. It's omitting it. Since we encode the field key with the length, when there is one, this is valid.